crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -0,0 +1,561 @@
+            Fee::Dynamic(fee) => fee.0,
+        };
+
+        let protocol_fee = order
+            .fee_policies
+            .iter()
+            .map(|fee_policy| {
+                match (fee_policy, order.side) {
+                    (
+                        FeePolicy::PriceImprovement {
+                            factor,
+                            max_volume_factor,
+                        },
+                        Side::Buy,
+                    ) => {
+                        let price_improvement_fee = price_improvement_fee_for_buy(
+                            order.sell.amount.0,
+                            order.buy.amount.0,
+                            executed.0,
+                            surplus_fee,
+                            uniform_sell_price,
+                            uniform_buy_price,
+                            *factor,
+                        )?;
+                        let max_volume_fee = volume_fee_for_buy(
+                            executed.0,
+                            surplus_fee,
+                            uniform_sell_price,
+                            uniform_buy_price,
+                            *max_volume_factor,
+                        )?;
+                        // take the smaller of the two
+                        Some(std::cmp::min(price_improvement_fee, max_volume_fee))
+                    }
+                    (
+                        FeePolicy::PriceImprovement {
+                            factor,
+                            max_volume_factor,
+                        },
+                        Side::Sell,
+                    ) => {
+                        let price_improvement_fee = price_improvement_fee_for_sell(
+                            order.sell.amount.0,
+                            order.buy.amount.0,
+                            executed.0,
+                            surplus_fee,
+                            uniform_sell_price,
+                            uniform_buy_price,
+                            *factor,
+                        )?;
+                        let max_volume_fee =
+                            volume_fee_for_sell(executed.0, surplus_fee, *max_volume_factor)?;
+                        // take the smaller of the two
+                        Some(std::cmp::min(price_improvement_fee, max_volume_fee))
+                    }
+                    (FeePolicy::Volume { factor }, Side::Buy) => volume_fee_for_buy(
+                        executed.0,
+                        surplus_fee,
+                        uniform_sell_price,
+                        uniform_buy_price,
+                        *factor,
+                    ),
+                    (FeePolicy::Volume { factor }, Side::Sell) => {
+                        volume_fee_for_sell(executed.0, surplus_fee, *factor)
+                    }
+                }
+            })
+            .fold(eth::U256::zero(), |acc, fee| match fee {
+                Some(fee) => acc + fee,
```

Comment:
If we use checked_mul do we also need to use checked_add ? I can see how this can become annoying when using iter functions instead of good old loops. Feel free to change back to whatever is easier (as long as the loop body is not overly complex)

Reply:
added checked function,